### PR TITLE
ARROW-108: [C++] Add Union implementation and IPC/JSON serialization tests

### PR DESCRIPTION
**Pull request by wesm**
Closes #206.

Still need to add test cases for JSON read/write and dense union IPC. Integration tests can happen in a subsequent PR (but the Java library does not support dense unions yet, so sparse only -- i.e. no offsets vector)

### CODE REVIEW

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -427,25 +426,54 @@ class ARROW_EXPORT StructArray : public Array {
...
 
 // ----------------------------------------------------------------------
 // Union
 
 class UnionArray : public Array {
+ public:
+  using TypeClass = UnionType;
+
+  UnionArray(const TypePtr& type, int32_t length,
+      std::vector<std::shared_ptr<Array>>& children,
```
Comment (xhochy) [inline cpp/src/arrow/array.h]: const?

File: cpp/src/arrow/array.h [hunk 2]
```diff
@@ -427,25 +426,54 @@ class ARROW_EXPORT StructArray : public Array {
...
+      const std::shared_ptr<Buffer>& type_ids,
+      const std::shared_ptr<Buffer>& offsets = nullptr, int32_t null_count = 0,
+      std::shared_ptr<Buffer> null_bitmap = nullptr);
```
Comment (xhochy) [inline cpp/src/arrow/array.h]: const ref

**Review by xhochy — CHANGES_REQUESTED** [review]
I see that you have changed  `ArrayPtr` in a lot but not all places. The changes adds more characters but in the end, it's more readable because it actually tells one that it is a `shared_ptr`. But then it would be better if we would get rid of it completely.

**Comment by wesm** [comment]
Tests are completed

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/array.h]: done

File: cpp/src/arrow/array.h (hunk 2 above)
Comment (wesm) [inline cpp/src/arrow/array.h]: done

**Comment by wesm** [comment]
Removed ArrayPtr completely

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**